libs/platform/ebpf_epoch.c

```diff
@@ -411,9 +369,7 @@ ebpf_epoch_exit()
     }
 
     if (is_preemptible) {
-        // Restore the thread's affinity mask.
-        ebpf_restore_current_thread_affinity(old_thread_affinity);
-        ebpf_semaphore_release(_ebpf_epoch_cpu_table[current_cpu].thread_entry_table_semaphore);
+        ebpf_semaphore_release(_ebpf_epoch_cpu_table[current_cpu].epoch_table_semaphore);
```

Comment:
nit: should we store `is_preemptible` when a thread calls `ebpf_epoch_enter` and assert that the current IRQL should still be the same when `ebpf_epoch_exit` is called?

This may avoid the following scquence:
1. Thread calls `ebpf_epoch_enter` at PASSIVE
2. Thread raises IRQL to DISPATCH
3. Thread calls `ebpf_epoch_exit`

Reply:
doesn't this scenario fall into the 'don't do that' category of windows driver programming best practices?  Rather than work around such (egregiously?) buggy code, what if we ensure that the entry-exit IRQLs match and bugcheck if they don't?

Reply:
Genuine question: How can the kernel thread raise its IRQL? This (thread) would be an epbf program, right? They can't call non-helper (i.e. OS kernel functions) functions, so the only way this could happen is if a helper function returned at an increased IRQL.  Is this possible/allowed or am I not seeing something?

Reply:
Ah. I'd forgotten about the IOCTL paths (among others) :-P  Ofc, anything is possible there. I see it now.  Thanks!

Reply:
Where are you proposing we track this? The epoch state currently is 64bits, so adding a flag will essentially double the epoch state size.

Reply:
Ideally it would have to be in epoch state only. One option is to have that flag under `#if debug`, so it is only present in debug builds, as anyways we only want to use it to catch issues in our tests.

Reply:
I am not sure it's really needed. Code that enters at passive and leaves at dispatch will be caught by driver verifier.

Reply:
hmm, with the below flow, DV will not catch it, but it will break epoch logic.

1. Thread calls ebpf_epoch_enter at PASSIVE
2. Thread raises IRQL to DISPATCH
3. Thread calls ebpf_epoch_exit
4. Thread lowers IRQL to PASSIVE

Reply:
Maybe file an issue for this?

Reply:
Done: #2270